Initial pass at webhook-subscription-2022 (#51)
solid/notifications: webhook-subscription-2022.md

# WebHookSubscription2022

## Abstract
The [Solid Notification Protocol](https://solid.github.io/notifications/protocol) defines a set of interaction patterns for agents to establish subscriptions to resources in a Solid Storage.

This specification defines a subscription type that applies these patterns to WebHooks.

## Status of This Document
TODO

## 1. Introduction
_This section is non-normative._

> The [Solid Notifications Protocol](https://solid.github.io/notifications/protocol) allows for many methods by which a client is notified of actions being performed on a resource. One such method of notification is webhooks. Webhooks allow servers to register a target URL with a Solid Pod. When that server needs to be notified of an action, that target URL will be called by the Solid Pod.

Reviewer: This is a non-normative section, so this is just a suggestion for consideration. Generally, with specification documents, the term "Solid Storage" is more precise than "Solid Pod". The informal "Solid Pod" is great in colloquial contexts, but for these documents, I think "Solid Storage" would be preferred.

Webhooks are an alternative to websockets that are more useful for server-based use cases. While a websocket solution would require a server to maintain a constant websocket connection with a Pod as long as it wants notifications, webhooks only require one registration.

> This document proposes an implementation for the `WebHookSubscription2022` Solid Notification protocol.

Reviewer: Is there a reason to change this type? It's fine to leave it as-is, if you'd prefer.

Author: I figured it's 2022, and this is a different version. There are servers floating around out there with that follow the old spec I wrote, so changing the version number would help us differentiate between them.

Reviewer 2: We agreed in the 2022-03-24 meeting to use `2021` instead of `2022` ( https://github.com/solid/notifications-panel/pull/52 ). I'd add that:

An Editor's Draft (without a "published" version floating around) is still at an early stage in the process. I think it'd do those servers a favour if they didn't need to carry the baggage around with 2021 if 2022 comes out. 2021 will have to be removed or renamed/updated into 2022 in any case. Sticking to 2021 will just mean that they'll update their current code and get on with the current state of the spec.

Reviewer 3: This comment is outdated, [562e888](https://github.com/solid/notifications/pull/51/commits/562e888a33681309c087acfb164702bbc3993a85) already reuses 2021

> ### 1.1. Specification Goals:

Reviewer: Should these move to the [use case document](https://solid.github.io/notifications-panel/notifications-ucr.html)?

Author: Didn't know that document existed. I'm fine with it moving, but does it make sense when these use cases are webhook-specific?

Reviewer 2: They can be a nice addition to the UCR document. I'm not interpreting those use cases as WebHook-specific. Besides... use cases shouldn't be solution-centric!

Reviewer 3: I think this can be handled in a separate PR, especially the UCR document is in another repo. Also, anyone can pick this as a task. This shouldn't slow down the merging of this PR.

 - **Subscribing Server**: A server interestest in a webhook alert. In this example it is "Liqid Chat," a chat app API hosted at `https://api.liqid.chat`.
 - **Solid Metadata Resouce**: A metadata resource compliant with the Solid Specification. In this example, it is hosted at `https://pod.example/.well-known/solid`.
 - **Subscription API**: an HTTP API at which a client can initiate a subscription to notification events for a particular set of resources. In this example, it is hosted at `https://pod.example/subscription`
 - **Authorization Server**: a Solid OIDC compliant identity server. In this example, it is hosted at `https://idp.example`
 - **Topic Resource**: A resource on a Pod that is being tracked for webhooks. In this example, it is hosted at `https://bob.pod.example/chat1.ttl`.
 - **Notification Server**: A server responsible for delivering a webhook notification. This may or may not be the same server as the resource server. In this example, the notification server is `https://pod.example/notifications/webhook/subscription`.

>  - **Notification Server WebId**: A notification server has its own webId known as the `Notification Server WebId`. In this example the Notification Server WebId is `https://pod.example/notifications/webhook/card.ttl#i`.

Reviewer: Minor: the canonical capitalization of this term is `WebID`

The following metadata document is returned:

```http
Content-Type: text/turtle

@prefix solid: <http://www.w3.org/ns/solid/terms#> .
@prefix notify: <http://www.w3.org/ns/solid/notifications#> .

<>
  a solid:StorageMetadata ;
  notify:hasNotificationChannel <#webhookNotification> .

<#webhookNotification>
  a notify:WebHookSubscription2022 ;
  notify:subscription <https://pod.example/notifications/webhook/subscription> ;
  notify:webid <https://pod.example/notifications/webhook/card.ttl#i> ;
  notify:features notify:rate, notify:expiration, notify:webhookAuth .
```

> Note the `nofify:subscription` and `notify:webid`. We'll be using those in future steps.

Reviewer: ```suggestion
Note the use of `nofify:subscription` and `notify:webid`. We'll be using those in future steps.
```

GET https://idp.example/jwks
```

#### 5. Receive capabilities

The auth server returns the JWKS.

If the token is not valid or the user does not have READ access to the resource, the Pod will return 403 to the server client.

#### 6. Receive subscription URL and metadata
The Pod returns a subscription URL to the subscribing server.

```http
Content-Type: application/ld+json

{
    "@context": "https://www.w3.org/ns/solid/notification/v1",
    "type": "WebHookSubscription2022",
    "target": "https://api.liqid.chat/webhook",
    "unsubscribe_endpoint": "https://pod.example/notifications/webhook/subscription/a59e24ba-8231-4b51-b60a-c0e04740f617"
}
```

Note the "unsubscribe_endpoint." We'll use that later when we unsubscribe from this hook.

At this point, the webhook has been successfully registered.

#### 7. Action happens on Topic Resource
Some action (either an update to the resource or a deletion of that resource) has taken place on the topic resource.

#### 8. Webhook Request (With token signed by the Pod Key)
A request is made to the subscribing server's registered webhook.

```http
POST https://api.liqid.chat/webhook
Authorization: DPoP <authToken>

> Dpop: <dpopToken>

Reviewer: ```suggestion
Dpop: <dpopProof>
```

   },
   "state": "1234-5678-90ab-cdef-12345678",
   "published":"2021-08-05T01:01:49.550044Z"
   "unsubscribe_endpoint": "https://pod.example/webhooks/subscription/a59e24ba-8231-4b51-b60a-c0e04740f617"
}
```

> The value of `<authToken>` is a DPoP bound JSON Web Token representing `https://pod.example/webhooks/card.ttl#i`. Each token corr

Reviewer: seems that this last sentence was cut off

> `<dpopToken>` is a dpop token with an `{ htu: "https://api.liqid.chat/webhook", htm: "POST" }`. Both tokens correspond to the authentication method outlined in Solid OIDC.

Reviewer: ```suggestion
`<dpopProof>` is a dpop proof containing the claims `{ htu: "https://api.liqid.chat/webhook", htm: "POST" }`. Both tokens correspond to the authentication method outlined in Solid OIDC.
```

#### 4.1.2 Delete

If a resource is deleted, the Pod `MUST` send a webhook request to all subscribed targets.

The webhook request `MUST` follow the standard outlined in the [Acivity Pub Delete Type](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-delete)

## 5. Unsubscribe API

All servers implementing the `WebHookSubscription2022` protocol `MUST` provide a unique resource for each subscription. This will be called the "Unsubscribe Endpoint."

If a `DELETE` request is received at the unsubscribe endpoint, the Pod `MUST` forget that subscription and any subsequent actions should not trigger a webhook request.

If a request is received at the unsubscribe endpoint and it does not include an Authorization or DPoP header, the Authorization or DPoP Headers are invalid, or the WebId in the authorization header does not correspond with the WebId that made the subscription, the Pod `MUST` reject the request.

## 6. Features
This section details the additional features built for `WebHookSubscription2021`.

### 6.1 webhook-auth
The `webhook-auth` feature allows subscribing servers to verify that a request came from a certain Pod.

> Pods that implement the `webhook-auth` feature `MUST` create an Access Token and DPoP token in accordance with the Solid OIDC specification and embed those tokens in the headers of all webhook requests.

Reviewer: ```suggestion
Pods that implement the `webhook-auth` feature `MUST` create an Access Token and DPoP proof in accordance with the Solid OIDC specification and include those tokens in the headers of all webhook requests.
```

## Further Considerations

>  - Using DPoP authentication in the webhook request does coincide with the common use case for Solid-OIDC, but it is overkill. There is no need for a security model that has two separate tokens (DPoP Token and the Auth token). Instead, we should consider using one token with the combined features (identity + htm and htu) of the two tokens.

Reviewer: Should this rather be brought up as an independent issue in the authN panel? It seems more like commentary and arguably a bit out of place in a specification document.

Author: Perhaps. The proper way to address this is also defining a non-DPoP version of Solid OIDC so a client can use that.

Reviewer 3: I think this section can stay and be used as a reference in the suggested AuthN panel issue. Once it gets addressed there this document can be updated.
